JP-3588: Use Pastasoss datamodel for NIRISS SOSS transform solution

tapastro opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3588](https://jira.stsci.edu/browse/JP-3588)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #8398 

<!-- describe the changes comprising this PR here -->
This PR modifies the algorithm for NIRISS SOSS data transforms, used to align the data with the expected spectral trace positions. Previously, the ATOCA algorithm derived a 3 parameter rotation/offset transform that struggled to deal with certain atypical scenarios; this has been removed, and the https://github.com/spacetelescope/pastasoss team has provided an adapted version of their algorithm for use in the pipeline.

The pastasoss algorithm has introduced a new reference file type, aptly named `pastasoss` holding a `PastasossModel`. These files are not yet on the CRDS OPS server - regression testing to be done using CRDS TEST for now, with the expectation that any okify'ed results will need updating once OPS is mirrored to TEST and/or OPS' CRDS server code is updated to handle the new reftype.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

tapastro commented:
Started a run against CRDS-TEST only for `test_niriss_soss.py`: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1717/

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+    return x_scaled
+
+
+def wavecal_model_order1_poly(refmodel, x, pwcpos):
```
add short docstring with parameter definitions, as it wasn't immediately obvious to me what `pwcpos` was at a glance

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+    """
+    This is the primary method for generate the gr700xd trace position given a
+    pupil wheel positions angle provided in the FITS header under keyword
+    PWCPOS. The traces for a given spectral order are found by perform a
```
```suggestion
    PWCPOS. The traces for a given spectral order are found by performing a
```

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+        order: str, subarray: str, interp: bool = True
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    This is the primary method for generate the gr700xd trace position given a
```
```suggestion
    This is the primary method for generating the gr700xd trace position given a
```

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    This is the primary method for generate the gr700xd trace position given a
+    pupil wheel positions angle provided in the FITS header under keyword
```
```suggestion
    pupil wheel position angle provided in the FITS header under keyword
```

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+    pupil wheel positions angle provided in the FITS header under keyword
+    PWCPOS. The traces for a given spectral order are found by perform a
+    rotation transformation using the refence trace positions at the commanded
+    PWCPOS=245.76 degrees. This methods yield sub-pixel performance and will be
```
```suggestion
    PWCPOS=245.76 degrees. This method yields sub-pixel performance and will be
```

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+        The pupil wheel positions angle provided in the FITS header under
+        keyword PWCPOS.
+    order : str, optional
+        The spectral order to compute the new traces for. Default is '12'.
```
```suggestion
        The spectral order for which to compute the new traces. Default is '12'.
```

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+        keyword PWCPOS.
+    order : str, optional
+        The spectral order to compute the new traces for. Default is '12'.
+        Support for order 3 will be added at a later date.
```
```suggestion
        Order 3 is currently unsupported
```

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+        # wavelength associated to trace at given pwcpos value
+        wavelengths = get_wavelengths(refmodel, x_new, pwcpos, int(order))
+
+        # return x_new, y_new, wavelengths
```
```suggestion
```

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+    ypad = np.copy(padding)
+
+    # No need to compute wavelengths across the entire detector, slightly larger than SUBSTRIP256 will do.
+    dimx, dimy = 2048, 300
```
consider moving hard-coded values into all-caps variable names at top of file

tapastro replied:
Did my best to move all hard-coded values to const vars.

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+    spectrace_1 = np.array([xtrace_order1, ytrace_order1, wave_grid])
+
+    # Set cutoff for order 2 where it runs off the detector
+    o2_cutoff = 1783
```
same here

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+    # Make wavemap from trace center wavelengths, padding to shape (296, 2088)
+    wavemin = 0.5
+    wavemax = 5.5
+    nwave = 5001
```
same here

emolter commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+    # Set cutoff for order 2 where it runs off the detector
+    o2_cutoff = 1783
+    w_o2_tmp = order2_wl[:o2_cutoff]
+    w_o2 = np.zeros(2040) * np.nan
```
where does the 2040 come from?

tapastro replied:
I believe this is the result of using the full frame, minus the 4 pixel reference pixel edges.

emolter commented on jwst/extract_1d/soss_extract/soss_extract.py:
```diff
@@ -508,24 +518,18 @@ def model_image(scidata_bkg, scierr, scimask, refmask, ref_files, box_weights, s
-    subarray : str
-        Subarray on which the data were recorded; one of 'SUBSTRIPT96',
-        'SUBSTRIP256' or 'FULL'.
-    transform : array or list, optional
```
Not sure the best place to put this comment, but it looks like `soss_kwargs["transform"]` is no longer used anywhere in the code.  Can `soss_transform` safely be removed from the arguments to `extract_1d_step`?  If so, don't forget to update the docs to reflect that.

tapastro replied:
I've attempted to fully excise this parameter from both code and the documentation.

melanieclarke commented on jwst/extract_1d/soss_extract/pastasoss.py:
```diff
@@ -0,0 +1,537 @@
+
+
+def get_wavelengths(
+    refmodel: PastasossModel, x: np.ndarray, pwcpos: float, order: int
```
For what it's worth, I vote for just removing all type hints - I've never found them more helpful than they are annoying to maintain.  But yes, we should file a ticket to sort it out either way; I can file it.

tapastro replied:
I agree that type hints without something like mypy/Pydantic for enforcement are not very helpful. They were a carryover and I'm happy to remove them.

tapastro commented on jwst/extract_1d/soss_extract/soss_extract.py:
```diff
@@ -38,39 +36,56 @@ def get_ref_file_args(ref_files, transform):
-        log.warning(msg_warning.format(2))
+    pastasoss_ref = ref_files['pastasoss']
+    if hasattr(pastasoss_ref.traces[0], "padding"):
+        pad = pastasoss_ref.traces[0].padding
```
Nice!

tapastro commented:
Requesting re-reviews after addressing comments!

On the topic of unit tests and documentation: I requested these of the IDT members when they supplied the code, and it never came to fruition. I think it's reasonable to request a paragraph or two of documentation from the NIRISS team, with appropriate links to the pastasoss repo and to the ATOCA paper: https://iopscience.iop.org/article/10.1088/1538-3873/ac8a77. I can file a ticket requesting this.

I have a feeling that NIRISS will not be interested in supplying unit tests, though we can ask.

emolter commented on jwst/extract_1d/soss_extract/soss_extract.py:
```diff
@@ -745,13 +743,16 @@ def model_image(scidata_bkg, scierr, scimask, refmask, ref_files, box_weights, s
         order_list = [1, 2]
     else:
-        order_list = [1, 2, 3]
+        # Regression in behavior? Used to run on three orders, but
+        # pastasoss does not support order 3
+        # order_list = [1, 2, 3]
+        order_list = [1, 2]
```
@tapastro it doesn't look like a change has been made in response to this comment

tapastro replied:
Hmm, not sure how I missed these. Latest commit should address these.

emolter commented on jwst/extract_1d/soss_extract/soss_extract.py:
```diff
@@ -1015,18 +1016,16 @@ def extract_image(decontaminated_data, scierr, scimask, box_weights, bad_pix='mo
     return fluxes, fluxerrs, npixels
```
@tapastro this one too, unless I missed it

tapastro replied:
^^

emolter reviewed (commented):
Still two unresolved comments from my last review.  Otherwise this looks good to me and I'll hit the approve button once those are commented on or otherwise resolved

melanieclarke reviewed (commented):
It looks like there's a failure in the unit tests - a doctest for `jwst.extract_1d.soss_extract.pastasoss.get_soss_traces`.

I know the team is keen to get this change in this build, so I don't want to hold it up here, but I agree with Ned that this change could use some unit tests and documentation.  I think some more code clean up will probably also be warranted, when the dust settles.  Can we file a ticket for follow up in the next build?

tapastro commented:
Yes, this bit of code is chock full of technical debt. I can file the ticket.

I had left some machinery in `get_soss_traces` that was left over from its use as a standalone repo, but I realize now that it no longer had any utility, so I've removed some of the offending code. I think it's in a functional state now? 😬

melanieclarke reviewed (approved):
Works for me, since the NIRISS SOSS team is happy with their testing.

emolter commented:
@tapastro just a note that we should follow up with [JP-3650](https://jira.stsci.edu/browse/JP-3650) once this is merged, perhaps by changing its status to "Ready for Testing" and re-assigning it to someone else?
